src/ru/nsu/ccfit/zuev/osu/OSUParser.java

```diff
@@ -427,18 +348,32 @@ private String getNextSectionName() {
     }
 
     private Map<String, String> parseSection() {
-        final Map<String, String> map = new HashMap<String, String>();
+        final Map<String, String> map = new HashMap<>();
 
         String s;
         try {
             while ((s = reader.readLine()) != null) {
-                if (s.matches(".+:.*") == false) {
+                // Remove leading and trailing whitespaces.
+                s = s.trim();
+
+                // Now that we've trimmed whitespaces, we can properly skip empty lines.
+                if (s.isEmpty()) {
+                    continue;
+                }
+
+                if (!s.matches(".+:.*")) {
                     if (s.matches("//.+")) {
                         continue;
                     } else {
+                        // If we don't meet a comment, it means we have arrived at a new section. In that case, go back to the previous line.
+                        // This ensures that reader.readLine() in getNextSectionName() returns the mentioned section.
```

Comment:
There is a leap in logic, "don't meet a valid key-value pair or comment" don't always means a new section in logic.
DIrectly check if this line is "[SectionName]" would be better.  

And if you have enough time, you can try to reorganize the .osu file parsing logic here, there are still some bad designed code.

Reply:
Ah, sure.